### ISSUE TITLE
Add reference image/grid option to resample function

### ISSUE THREAD

**Issue by nwschurink** [fepegar/torchio#118]
**Is your feature request related to a problem? Please describe.**
In my dataset I have images with different voxelgrids/spacings. As a result if e.g. a `mask` and `T2` image get resampled to 1x1x1mm voxels, it is possible that the physical locations of these voxel do not match. This happens for example when a mask is made on a different MRI sequence with a lower resolution, such as a `DWI` scan.

It would be nice if there is an option to resample all images to 1 specific reference image. (e.g. the `T2` voxelgrid)

**Describe the solution you'd like**
When resampling using ITK you can set a reference image, such that the voxelgrids overlap.

```
resampler = sitk.ResampleImageFilter()
resampler.SetInterpolator(sitk.sitkLinear)
resampler.SetReferenceImage(reference_image)
mask_image_resampled_to_reference_spacing = resampler.Execute(mask_image)
```

**Describe alternatives you've considered**
Pre-processing all my data outside torchio, however this would probably be a nice feature to add.

**Additional context**
I could take a look at how to implement this into the existing code. I'm not sure whether you would want such an option in the  `torchio.transforms.Resample` function (e.g. with something like a `reference_image='T2'` argument), or if this would be better implemented in a seperate transform.

**Comment by romainVala** [comment]
I like the proposition to add a target space for the resample (I do it with nibabel, with resample_from_to) but I do not understand why there is a mismatch
I can imagine this mismatch happen since you have different field of view
using a composition of resampled to 1mm and crop to get the same field of view does not solve it ?

Never the less I thing it is a good idea to allow a reference space

**Comment by fepegar** [comment]
I think what @nwschurink has two advantages:
1. Transform is faster
1. Information is not lost at interpolation

This can be applied to a sample containing e.g. a T1 and a T2 that are aligned but not in the same grid. The transform can be `Resample(target='T1')`, and the resampling strategy will be chosen according to the type of `target`.

**Comment by nwschurink** [comment]
Yes, this happens when the T1 and T2 are aligned but not in the same grid. The patient coordinatesystem is the same for both scans, however the voxelgrids are different. 

There are many cases when this could happen, e.g. different voxelsizes, slice direction, field of view etc. 

For my own pipeline, what I usally do is resample 1 image modality to a certain voxelsize. e.g. resample T2 to 1x1x1mm voxels and then use the sitk.ResampleImageFilter code above to resample all other modalities to the same voxelsize and grid by setting the resampled T2 image as the reference image.

If the 1x1x1 resampling of the reference image is not needed, this would indeed make the processing a little bit faster.

**Comment by nwschurink** [comment]
A note that could be added to the docs is that: Setting a reference image space would also allow using the CropOrPad function when the shape of the Image and Mask are inconsistent (i.e. in the case that the Mask image only contains the bounding box of the mask).

Using the Image as target space would automatically resample the Mask to have the same data dimensions

**Comment by fepegar** [comment]
Yeah I was thinking that maybe it'd be nice to ignore the mask shape when checking consistency in CropOrPad with a mask and automatically resampling the mask in the space of the other images. Then I thought that might be a bit too much magic preprocessing, but I guess that's one of the points of this library: taking the pain out of these little things.

**Comment by nwschurink** [comment]
Yea that would be nice idea, to let the mask be automatically resampled. I think this could be the default option when there is only 1 Image and 1 Mask present in a subject. :)

I was thinking though, you would need to specifically address resampling to a common spacing in the case that you have multiple Images with different spacings. For example when you have these images:

```
{'adc': torch.Size([1, 192, 156, 36]),
 'bhigh': torch.Size([1, 192, 156, 36]),
 'mask': torch.Size([1, 18, 22, 6]),
 't2': torch.Size([1, 320, 320, 36])}
```
Then it is impossible to decide this without input of the user right? So the target argument would be needed anyway. Would it make any sense to add this to the CropOrPad function as well? Or do you think it would be better to then mention in the Error message + docs to first use Resample with a target spacing and _then_ use the crop or pad function

**Comment by fepegar** [comment]
Resolved in #121.
